### PR TITLE
[automation] Added API for ScriptEngineFactory implementations to pull presets

### PR DESCRIPTION
**Pull request by jpg0**
Existing API only allows engine factory implementors to request presets are pushed into existing scopes, rather than returning them directory, so that they can be bound into module or library systems. This change allows implementors to capture the presets requested (and potentially ignore any pushed), and therefore expose them via idiomatic import mechanisms for the language being implemented.

Specifically, I am using this for Javascript CommonJS module support. Without this change it's possible, but ugly (using reflection) and tricky (using weakrefs to ensure engines can be collected). This change is minimal and backwards compatible with the current API.

Signed-off-by: Jonathan Gilbert <jpg@trillica.com>

### CODE REVIEW

File: bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptExtensionManagerWrapper.java [hunk 1]
```diff
@@ -56,7 +57,7 @@ public Object get(String type) {
         return manager.getDefaultPresets();
     }
 
-    public void importPreset(String preset) {
-        manager.importPreset(preset, container.getFactory(), container.getScriptEngine(), container.getIdentifier());
+    public Map<String, Object> importPreset(String preset) {
```
Comment (cweitkamp) [inline bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptExtensionManagerWrapper.java]: I know that you are not the initial contributor of these classes but may I ask you to add some JavaDoc to these methods?

Comment (jpg0) [reply bundles/org.openhab.core.automation.module.script/src/main/java/org/openhab/core/automation/module/script/internal/ScriptExtensionManagerWrapper.java]: I've added docs to the this method. TBH I'm not so sure about the other methods as I haven't used them. I know that most of the actual implementation is in `ScriptExtensionManager`, which unfortunately is also missing docs.

**Comment by openhab-5iver** [comment]
Please do not merge this! It requires extensive testing and I very much question the need for these changes. I would very much like to have the time to review this PR myself, but my time is very limited. I am still reviewing the GraalJS addon.

**Comment by jpg0** [comment]
@openhab-5iver have you looked at the code changes in this PR? It's got to be one of the most benign changes there is. Why do you say that it needs extensive testing? All it does is return a collection of items that is also pushed elsewhere. There is next to no logic involved, and nothing (yet) actually uses the returned objects.

I don't agree that it means the docs would require changes; existing code can operate exactly as before. I do agree that it introduces an extra feature which we could choose to document (`importPreset` returns a map of objects), but personally I don't think that we should, as I don't think that we have agreed that it should be used that way, nor do either of us have any plans to use it that way.

**Comment by cweitkamp** [comment]
@openhab-5iver I did not test anything. Just code review.

**Comment by openhab-5iver** [comment]
> It's got to be one of the most benign changes there is. 

There is nothing benign in automation, especially scripted automation! Small changes can have huge effects. If the maintainers don't merge them before I get a chance, I have been doing extensive functional testing on all automation PRs.

>  but personally I don't think that we should, as I don't think that we have agreed that it should be used that way, nor do either of us have any plans to use it that way.

And that is exactly my hesitation about this PR, until I have a better understanding of what it does, what it will be used for, and what impacts there are to existing functionality. 

If new features are being added, the documentation should definitely be updated.

**Comment by cweitkamp** [comment]
@jpg0 @openhab-5iver What is the status of your automation PRs. We would like to start working on #1290 which might result in some merge conflicts. Can we finalize some of them before?

**Comment by openhab-5iver** [comment]
#1253 and https://github.com/openhab/openhab-addons/pull/6399 (which I have other concerns with) are hung up until a decision is made about where ScriptEngineFactories should live, but there is a deeper question about splitting automation from core. I plan to submit an issue for that over the holidays. 

I made a comment in #1196. 

#1256 and #1264 are tailored for the graaljs ScripEngineFactory and require extensive testing. Jonathan has not reported what, if anything, has been tested. I have not tested these for any language. This code is fragile and I believe it is not well known by any maintainers, so I advice being very cautious with any changes to this code. These PRs are only introducing functionality that would be useful with Jonathan's implementation of the graaljs ScriptEngineFactory. It would be best to address that PR first before these. If I can keep out of the eggnog, I hope to get to them this year.

**Comment by jpg0** [comment]
> #1256 and #1264 are tailored for the graaljs ScripEngineFactory

> These PRs are only introducing functionality that would be useful with Jonathan's implementation of the graaljs ScriptEngineFactory.

These were explicitly submitted independently because they target any language, not only graaljs. The first allows any script to _get_ a preset (rather than requesting it be injected) and the second provides an API to listen for scripts being unloaded. They are in no way related to graal or javascript.

**Pull request merged**